pkg/scalers/aws/aws_config_cache.go

```diff
@@ -0,0 +1,131 @@
+	"context"
+	"encoding/hex"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/credentials"
+	"github.com/aws/aws-sdk-go-v2/credentials/stscreds"
+	"github.com/aws/aws-sdk-go-v2/service/sts"
+	"github.com/go-logr/logr"
+	"golang.org/x/crypto/sha3"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+type cacheEntry struct {
+	config *aws.Config
+	usages map[string]bool // Tracks the resources which have requested the cache
+}
+type sharedConfigCache struct {
+	sync.Mutex
+	items  map[string]cacheEntry
+	logger logr.Logger
+}
+
+func newSharedConfigsCache() sharedConfigCache {
+	return sharedConfigCache{items: map[string]cacheEntry{}, logger: logf.Log.WithName("aws_credentials_cache")}
+}
+
+func (a *sharedConfigCache) getCacheKey(awsAuthorization AuthorizationMetadata) string {
+	key := "keda"
+	if awsAuthorization.AwsAccessKeyID != "" {
+		key = fmt.Sprintf("%s-%s-%s", awsAuthorization.AwsAccessKeyID, awsAuthorization.AwsSecretAccessKey, awsAuthorization.AwsSessionToken)
+	} else if awsAuthorization.AwsRoleArn != "" {
+		key = awsAuthorization.AwsRoleArn
+	}
+	// to avoid sensitive data as key and to use a constant key size,
+	// we hash the key with sha3
+	hash := sha3.Sum224([]byte(key))
+	return hex.EncodeToString(hash[:])
+}
+
+func (a *sharedConfigCache) GetCredentials(ctx context.Context, awsRegion string, awsAuthorization AuthorizationMetadata) (*aws.Config, error) {
+	a.Lock()
+	defer a.Unlock()
+	key := a.getCacheKey(awsAuthorization)
+	if cachedEntry, exists := a.items[key]; exists {
+		cachedEntry.usages[awsAuthorization.ScalerUniqueKey] = true
+		a.items[key] = cachedEntry
+		return cachedEntry.config, nil
+	}
+
+	configOptions := make([]func(*config.LoadOptions) error, 0)
+	configOptions = append(configOptions, config.WithRegion(awsRegion))
+	cfg, err := config.LoadDefaultConfig(ctx, configOptions...)
+	if err != nil {
+		return nil, err
+	}
+
+	if awsAuthorization.UsingPodIdentity {
+		if awsAuthorization.AwsRoleArn != "" {
+			cfg.Credentials = a.retrievePodIdentityCredentials(cfg, awsAuthorization.AwsRoleArn)
+		}
+	} else {
+		cfg.Credentials = a.retrieveStaticCredentials(awsAuthorization)
+	}
+
+	newCacheEntry := cacheEntry{
+		config: &cfg,
+		usages: map[string]bool{
+			awsAuthorization.ScalerUniqueKey: true,
+		},
+	}
+	a.items[key] = newCacheEntry
+
+	return &cfg, nil
+}
+
+func (a *sharedConfigCache) retrievePodIdentityCredentials(cfg aws.Config, roleArn string) *aws.CredentialsCache {
+	stsSvc := sts.NewFromConfig(cfg)
+	webIdentityTokenFile := os.Getenv("AWS_WEB_IDENTITY_TOKEN_FILE")
+
+	webIdentityCredentialProvider := stscreds.NewWebIdentityRoleProvider(stsSvc, roleArn, stscreds.IdentityTokenFile(webIdentityTokenFile), func(options *stscreds.WebIdentityRoleOptions) {
+		options.RoleSessionName = "KEDA"
+	})
+	var cachedProvider *aws.CredentialsCache
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
```

Comment:
We shouldn't call `context.Background()` here, but get one from `GetCredentials()`